pageserver/src/repository.rs

```diff
@@ -167,15 +217,159 @@ pub trait Timeline {
+            }
+            if key.tag.rel != prev_tag && key.lsn <= lsn {
+                prev_tag = key.tag.rel;
+                dbs.push(prev_tag); // collect unique tags
+            }
+            iter.next();
+        }
+        return Ok(dbs);
+    }
 
     /// Get vector of prepared twophase transactions
-    fn get_twophase(&self, lsn: Lsn) -> Result<Vec<TransactionId>>;
+    fn get_twophase(&self, lsn: Lsn) -> Result<Vec<TransactionId>> {
+        let key = RepositoryKey {
+            // minimal key
+            tag: BufferTag {
+                rel: RelTag {
+                    forknum: pg_constants::PG_TWOPHASE_FORKNUM,
+                    spcnode: 0,
+                    dbnode: 0,
+                    relnode: 0,
+                },
+                blknum: 0,
+            },
+            lsn: Lsn(0),
+        };
+        let mut gxacts = Vec::new();
+
+        let mut iter = self.iterator();
+        iter.first(&key);
+        while iter.valid() {
+            let key = iter.key();
+            if key.tag.rel.forknum != pg_constants::PG_TWOPHASE_FORKNUM {
+                break; // we are done with this fork
+            }
+            if key.lsn <= lsn {
+                let xid = key.tag.blknum;
+                let tag = BufferTag {
+                    rel: RelTag {
+                        forknum: pg_constants::PG_XACT_FORKNUM,
+                        spcnode: 0,
+                        dbnode: 0,
+                        relnode: 0,
+                    },
+                    blknum: xid / pg_constants::CLOG_XACTS_PER_PAGE,
+                };
+                let clog_page = self.get_page_at_lsn(tag, lsn)?;
+                let status = transaction_id_get_status(xid, &clog_page[..]);
+                if status == pg_constants::TRANSACTION_STATUS_IN_PROGRESS {
+                    gxacts.push(xid);
+                }
+            }
+            iter.next();
+        }
+        return Ok(gxacts);
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct RepositoryKey {
+    pub tag: BufferTag,
+    pub lsn: Lsn,
+}
+
+impl RepositoryKey {
+    fn pack(&self, buf: &mut BytesMut) {
+        self.tag.pack(buf);
+        buf.put_u64(self.lsn.0);
+    }
+    fn unpack(buf: &mut Bytes) -> RepositoryKey {
+        RepositoryKey {
+            tag: BufferTag::unpack(buf),
+            lsn: Lsn::from(buf.get_u64()),
+        }
+    }
+
+    fn from_slice(slice: &[u8]) -> Self {
+        let mut buf = Bytes::copy_from_slice(slice);
+        Self::unpack(&mut buf)
+    }
+
+    fn to_bytes(&self) -> BytesMut {
+        let mut buf = BytesMut::new();
+        self.pack(&mut buf);
+        buf
+    }
+}
+
+pub trait RepositoryIterator {
+    fn first(&mut self, key: &RepositoryKey);
+    fn last(&mut self, key: &RepositoryKey);
+    fn next(&mut self);
+    fn prev(&mut self);
+    fn valid(&self) -> bool;
+    fn key(&self) -> RepositoryKey;
+    fn value(&self) -> &[u8];
```

Comment:
In rust all iterators should impl the `Iterator` trait. Otherwise they're just going to be confusing, and won't work in the expected ways, e.g. "for x in my_iter { ... }" assumes `my_iter` is `Iterator`.

Reply:
I first try to implement Rust interator interface, but then I think that there is not so much sense in it. We are not going to provide to the users some abstraction API for adding custom storage. Actually I do not think that we will have many storage implementation at all. Currently storage abstraction is introduced to make it possible to have RocksDB and our own implementation. In the future I think RocksDB will be removed.

So I think it will be mire important to make iterator more efficient and easy to use than conforming to Rist standard iterator API. As far as RocksDB is the only implementation avaiable at this moment, I really make it similar with RocksDB.  And it is not only because ot RawIterator impkementaiton in RocksDB, but also because of the way it is used in higher level code (in most cases it is not needed iteration trough all records: for x  in ..., but most often it is used to perform serve in different directions, so using first/last operation is more convenient and afficient than creating two different iterators).

If for our own storage will we prefer to have some different API, it can be changed.

Reply:
I don't think this is a good reason to avoid `Iterator`. The reason to use `Iterator` is so that other people can easily understand what your code does, and for compatibility with every other part of the std library and all third-party crates.

I don't think that this code is more efficient. It's just more confusing. At best, it's premature optimization.

This makes things difficult for everyone else who is familiar with Rust and expects things to be done the standard way.

I would suggest learning Rust for more time, and using the std library traits, interfaces, and calling conventions, before deciding to ignore the standards.

Reply:
First of all we need not just standard iterator which iterates all collection records in unspecified order. We need bidirectional iterator with possibility to specify starting point (inclusive or exclusive).

Well, it doesn't mean that it is not possible to achieve it implementing standard Iterator trait. But frankly speaking I do no think that it will somehow simplify usage of this iterator or make this code more understandable.

Concerning efficiency: the proposed iterator is certainly more efficient because it allows to eliminate extra dynamic memory allocation (one per each item). Is it critical? I do no think so. But why RocksDB Rust API includes both standard Iterator and RawIterator (which API is reused)? Also this option-oriented programming... To extract next key-value pair I have to do three unwraps. Why do I need to do it if I know for sure that in my case both key and value are always defined?

If we are going to publish storage API, then I 100% agree with you that our storage iterator should implement standard Iterator treat. But if we are using it only internally and may be only temporary, I do not see much sense in using less convenient and efficient but more standard abstraction.

In any case, I do think that iterator API is so important. Much more iterating question is how to address data in the storage (key format) and should we have single key-value storage or separate relational and non-relational data.

Reply:
I agree that the rocksdb "iterator" types are bad. The "raw" iterators map directly to the [C++ library interface](https://github.com/facebook/rocksdb/blob/master/include/rocksdb/iterator.h), and don't map easily to the lifetimes one expects from a Rust `Iterator`.

It would be better to conceal the rocksdb implementation details, and not let them pollute the `Timeline` interface.